lib/Basic/RedirectedProcess.cpp

```diff
@@ -0,0 +1,217 @@
+  innerProcessInfo->writeHandle = pipeHandles[1];
+
+  return true;
+}
+
+bool RedirectedProcess::execute(const char *path, bool setGroupFlags,
+                                char *const *args, char *const *envp,
+                                std::mutex &spawnedProcessesMutex) {
+  // Initialize the spawn attributes.
+  posix_spawnattr_t attributes;
+  posix_spawnattr_init(&attributes);
+
+  // Unmask all signals
+  sigset_t noSignals;
+  sigemptyset(&noSignals);
+  posix_spawnattr_setsigmask(&attributes, &noSignals);
+
+// Reset all signals to default behavior.
+//
+// On Linux, this can only be used to reset signals that are legal to
+// modify, so we have to take care about the set we use.
+#if defined(__linux__)
+  sigset_t mostSignals;
+  sigemptyset(&mostSignals);
+  for (int i = 1; i < SIGUNUSED; ++i) {
+    if (i == SIGKILL || i == SIGSTOP)
+      continue;
+    sigaddset(&mostSignals, i);
+  }
+  posix_spawnattr_setsigdefault(&attributes, &mostSignals);
+#else
+  sigset_t mostSignals;
+  sigfillset(&mostSignals);
+  sigdelset(&mostSignals, SIGKILL);
+  sigdelset(&mostSignals, SIGSTOP);
+  posix_spawnattr_setsigdefault(&attributes, &mostSignals);
+#endif
+
+  // Establish a separate process group.
+  posix_spawnattr_setpgroup(&attributes, 0);
+
+  // Set the attribute flags.
+  unsigned flags = POSIX_SPAWN_SETSIGMASK | POSIX_SPAWN_SETSIGDEF;
+  if (setGroupFlags)
+    flags |= POSIX_SPAWN_SETPGROUP;
+
+// Close all other files by default.
+//
+// FIXME: This is an Apple-specific extension, and we will have to do
+// something else on other platforms (and unfortunately, there isn't
+// really an easy answer other than using a stub executable).
+#ifdef __APPLE__
+  flags |= POSIX_SPAWN_CLOEXEC_DEFAULT;
+#endif
+
+  posix_spawnattr_setflags(&attributes, flags);
+
+  // Setup the file actions.
+  posix_spawn_file_actions_t fileActions;
+  posix_spawn_file_actions_init(&fileActions);
+
+  // Open /dev/null as stdin.
+  posix_spawn_file_actions_addopen(&fileActions, 0, "/dev/null", O_RDONLY, 0);
+
+  // Create a pipe to use to read the command output, if necessary.
+  if (shouldCaptureOutput) {
+    // Open the write end of the pipe as stdout and stderr.
+    posix_spawn_file_actions_adddup2(&fileActions, innerProcessInfo->writeHandle, 1);
+    posix_spawn_file_actions_adddup2(&fileActions, innerProcessInfo->writeHandle, 2);
+
+    // Close the read and write ends of the pipe.
+    posix_spawn_file_actions_addclose(&fileActions, innerProcessInfo->readHandle);
+    posix_spawn_file_actions_addclose(&fileActions, innerProcessInfo->writeHandle);
+  } else {
+    posix_spawn_file_actions_adddup2(&fileActions, 1, 1);
+    posix_spawn_file_actions_adddup2(&fileActions, 2, 2);
+  }
+
+  // Spawn the command.
+  pid_t pid;
+  {
+    // We need to hold the spawn processes lock when we spawn, to ensure that
+    // we don't create a process in between when we are cancelled.
+    std::lock_guard<std::mutex> guard(spawnedProcessesMutex);
```

Comment:
API simplification: If you make this class only able to launch one process *and* move the initialization out *and* make just one method which does the actual spawn, then the client could hold the lock and not expose it through this API.